### PR TITLE
initial attempt to enable client/server annotations anywhere

### PR DESCRIPTION
**Pull request by jamescheney**
work in progress on #907

### CODE REVIEW

**Comment by jamescheney** [comment]
It sounded like people were happy with merging, perhaps a quick review would be helpful.  In particular we are currently not using the new features in any tests whatsoever (even just syntax/typechecking) so suggestions what tests should be added are welcome.

**Comment by dhil** [comment]
Yes! I will have a look tomorrow!

File: core/evalir.ml [hunk 1]
```diff
@@ -245,11 +245,12 @@ struct
...
-      let env =
+      let env, xs =
         match z, fvs with
-        | Some z, Some fvs -> Value.Env.bind z (fvs, Scope.Local) env
-        | None, None -> env
-        | _, _ -> assert false
+        | Some z, Some fvs -> Value.Env.bind z (fvs, Scope.Local) env, xs
+        | None, None -> env, xs
+        | Some z, None -> env, z::xs (* kludge to deal with possible mismatch *)
+        | None, Some _ -> assert false
```
Comment (jamescheney) [inline core/evalir.ml:253]: This was an attempt to work around an apparent bug in closure conversion which surfaces if you have local client functions with non-empty closures.  However I'm, sure it is not the right way to fix this problem, so I will undo it.  The right thing to do is merge the rest of this PR and then make a new issue describing the bug that needs to be worked around.

**Review by dhil — APPROVED** [review]
LGTM.

**Pull request merged**